pilot/pkg/proxy/envoy/v2/datastructures.go

```diff
@@ -0,0 +1,139 @@
+// 4. When an endpoint is deleted, it needs to be removed from all associated clusters.
+// 5. When one or more labels are added to an endpoint, the label selectors from clusters
+//    that "do not have this endpoint" need to be reevaluated to see if the newly added labels
+//    would cause this endpoint to be added to these clusters.
+// 6. When one or more labels are "removed" from an endpoint, the clusters currently associated
+//    with the endpoint need to be reevaluated to see if the endpoint still qualifies to be a member
+//    of these clusters.
+
+// Clusters (react to updates in services, dest rules, ext services)
+// 1. Every service (registry defined, external) in the system has a corresponding cluster
+// 2. Every subset expressed in a DestinationRule will have an associated cluster.
+// 3. Clusters are added/removed when a service is added/removed or when subsets are
+//    added/removed from DestinationRule or when a service is updated (ports are added/removed)
+// 4. Clusters need to be updated when new settings (e.g., circuit breakers, auth) are added
+//    via DestinationRules.
+
+// Routes (react to updates in services, route rules)
+// 1. Envoy RDS response is essentially an array of virtual hosts on a port or TCP routes on a port.
+// 2. Every HTTP service in the system will have a virtual host and atleast a default route (/)
+// 3. The HTTP service could have additional routes based on route rules.
+// 4. Every TCP service in the system will have a TCPRoute (or a filter chain match) associated
+//    with its service IP if present.
+// 5. If a TCP service has no service IP, it will have a listener on the service port and a
+//    catch all TCPRoute
+// 6. Every HTTP port of a service will share the same route configuration (list of virtual hosts)
+//    until further delineation is created by route rules.
+// 7. Similarly, every TCP port of a service will share the same TCPRoute.
+// 7. Routes change (add/delete/update) typically when route rules change
+// 9. Routes can also change when ports are added/removed from the service
+
+// Listeners (react to updates in services, gateways, ext services)
+// 1. Every unique port in the system will have an associated listener.
+// 2. A service can generate one or more listeners based on the number of ports it listens on
+// 3. New listeners are added when a service (incl gateway/external service) with unique
+//    (previously unused) set of ports are added.
+// 4. Listeners are updated when one or more services/gateways/extsvs are added/removed to/from
+//    the same listener port.
+// 5. Listeners are removed when there are no services/gateways/extsvcs listening on the listener
+//    port.
+
+// The indices here are doubly linked. A pair of objects are dependent on each other if they have
+// to be updated when one changes.
+
+// BoxedVirtualHost holds the virtual host proto representation and pointers to
+// owning route configurations.
+type BoxedVirtualHost struct {
```

Comment:
'boxed' is another term that may be confusing - but may be fine.

Reply:
I am open for a better word personally, but couldn't find one.. Essentially, this is a wrapper over the proto struct, wherein I am just adding some pointers to other related things.